### PR TITLE
Adding Firebase Auth

### PR DESCRIPTION
**Pull request by KevinSchildhorn**
Adding support for Firebase Authentication

Adds a Login/Logout button in the settings screen, that will authenticate with google sign-in using Firebase. Then the user state is sent to the common code, to be used later.

## Shared Code
* Adding `UserContext` and  `UserData` for commonly saved user information
* Currently passing this information to the `UserIdProvider` for now, will be handled later
* Adding abstract `AuthenticationService` for calling for sign in / sign out from common ui to platform specific code
* Passing `AuthenticationService` into `SettingsViewModel` and using that for compose calls

## Platform Code
* Adding new Id "co.touchlab.droidconauthtest" for testing
* Adding `FirebaseAuth` to android and iOS using Gradle and CocoaPods
* Adding `FirebaseService` which is implementation of `AuthenticationService` for each platform, to handle authentication
* Adding authentication listener in `MainActivity` and `AppDelegate` to save changes to the common code
* Adding the ClientID as a `secret` in Android. Not sure if this is needed

## TODO
We are unsure about saving the clientID in public or as a secret. In android it is already a secret, however in iOS it's added in a URLScheme so it's more challenging to keep secret. Will look into this more.

Simulator Screen Recording - iPhone 15 Pro - 2024-03-28 at 12 29 38

Screen_recording_20240328_124442.webm

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import co.touchlab.droidcon.UserContext
+import co.touchlab.droidcon.UserData
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.update
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+abstract class AuthenticationService(isSignedIn: Boolean) : KoinComponent {
```
Comment (samhill303) [inline shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt:11]: Could it be possible to fit this a little better into the exiting architecture? If we remove the googleSignIn function from this class then I think we could make this class an interface with a default implementation in the `impl` folder like the other services here.

googleSignIn could then be it's own service interface, but it's implementation would live in the app layer (following the example of AnalyticsService)

File: android/build.gradle.kts [hunk 2]
```diff
@@ -21,19 +21,23 @@ android {
...
-        applicationId = "co.touchlab.droidcon.london"
+        applicationId = "co.touchlab.droidconauthtest"
         minSdk = libs.versions.minSdk.get().toInt()
         targetSdk = libs.versions.targetSdk.get().toInt()
         versionCode = 20201
         versionName = "2.2.0"
         testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+
+        val clientId = properties.getProperty("clientId", "")
+        buildConfigField("String", "CLIENT_ID", clientId)
```
Comment (samhill303) [inline android/build.gradle.kts:32]: Can you update the readme with instructions on how to set your clientID for both android and iOS?

File: shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt (hunk 1 above)
Comment (KevinSchildhorn) [reply shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt]: Ok, I was able to separate into an `AuthenticationService` and a `GoogleSignInService`, hopefully this matches the architecture a bit more

File: android/build.gradle.kts (hunk 2 above)
Comment (KevinSchildhorn) [reply android/build.gradle.kts:32]: Updated the readme

**Comment by KevinSchildhorn** [comment]
Updated buttons

| Signed Out | Signed In |
| ----------- | ---------- | 
| Screenshot_20240329_124907 | Screenshot_20240329_124931 |
| Simulator Screenshot - iPhone 15 Pro - 2024-03-29 at 12 47 41 | Simulator Screenshot - iPhone 15 Pro - 2024-03-29 at 12 53 53 |

File: shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt (hunk 1 above)
Comment (samhill303) [reply shared/src/commonMain/kotlin/co/touchlab/droidcon/domain/service/AuthenticationService.kt]: looks great!

**Comment by KevinSchildhorn** [comment]
Updated style

|  | signed in | signed out |
| - | -------- | --- |
| android | Screenshot_20240401_143023 | Screenshot_20240401_143031 |
| ios | Simulator Screenshot - iPhone 15 Pro - 2024-04-01 at 15 02 59 | Simulator Screenshot - iPhone 15 Pro - 2024-04-01 at 15 03 02 |

**Comment by KevinSchildhorn** [comment]
Updated style

|  | signed in | signed out |
| - | -------- | --- |
| android | Screenshot_20240402_105704 | Screenshot_20240402_105712 |
| ios | Simulator Screenshot - iPhone 15 Pro - 2024-04-02 at 11 13 28 | Simulator Screenshot - iPhone 15 Pro - 2024-04-02 at 11 06 57 |

**Pull request closed without merging**